### PR TITLE
fix resume fsdp

### LINKED ISSUE #23034: Cannot resume FSDP optimizer state
This line does not save optimizer state correctly when using FSDP.

https://github.com/huggingface/transformers/blob/88399476c3892435395618ed37993176dbb0de73/src/transformers/trainer.py#L2383

It should use FSDP's full_optim_state_dict to collect optimizer states from different processes.
```python
FSDP.full_optim_state_dict(self.model, self.optimizer)
```

### PR DESCRIPTION
**Pull request by qywu**
# What does this PR do?

Fixes [# 23034](https://github.com/huggingface/transformers/issues/23034)

When training a model with FSDP, the checkpoint is not saved and loaded correctly. Only rank 0's optimizer state dict is saved. This PR fixes this issue.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@pacman100

### CODE REVIEW

**Review by pacman100 — APPROVED** [review]
Thank you @qywu for the super quick fix, LGTM! 🤗

**Review by pacman100 — CHANGES_REQUESTED** [review]
Hello, I went over it again and noticed that you aren't saving and loading the optimizer state only on rank 0. Please do that. Refer the implementation in accelerate here for reference: https://github.com/huggingface/accelerate/blob/main/src/accelerate/utils/dataclasses.py#L924-L952

**Comment by qywu** [comment]
I have fixed the issues. The optimizer saving had no problems. For using [scatter_full_optim_state_dict](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.scatter_full_optim_state_dict), indeed loading on rank 0 is enough can save CPU memory.

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2388,7 +2394,11 @@ def _save_checkpoint(self, model, trial, metrics=None):
                     torch.save(self.scaler.state_dict(), os.path.join(output_dir, SCALER_NAME))
         elif self.args.should_save and not self.deepspeed:
             # deepspeed.save_checkpoint above saves model/optim/sched
-            torch.save(self.optimizer.state_dict(), os.path.join(output_dir, OPTIMIZER_NAME))
+            if self.fsdp:
+                torch.save(full_osd, os.path.join(output_dir, OPTIMIZER_NAME))
```
Comment (pacman100) [inline src/transformers/trainer.py:2398]: saving on rank 0 should be efficient and enough, right?

Comment (qywu) [reply src/transformers/trainer.py:2398]: I believe `self.args.should_save` in this case is already handling saving on rank 0

Comment (pacman100) [reply src/transformers/trainer.py:2398]: oh, okay, got it, thank you!

**Review by pacman100 — APPROVED** [review]
Thank you @qywu for iterating 🤗

**Comment by pacman100** [comment]
cc @sgugger for a second look

**Review by sgugger — APPROVED** [review]
Thanks for the fix!

**Pull request merged**